awx_collection/plugins/action/generate_mesh.py

```diff
@@ -0,0 +1,157 @@
+            res[control_node] = data[type]["hosts"][(index + 1) :]
+        return res
+
+    def generate_topology_from_input(type=None, data=None):
+        res = {}
+        if type is None:
+            return None
+        if type not in data.keys():
+            return None
+        for node in data[type]["hosts"]:
+            # check to see if vertex exists
+            if not node in res:
+                res[node] = []
+            if "peers" in (data["_meta"]["hostvars"][node].keys()):
+                for peer in (data["_meta"]["hostvars"][node]["peers"]).split(","):
+                    # handle groups
+                    if peer in data.keys():
+                        ## list comprehension to produce peers list. excludes circular reference to node
+                        res[node] = res[node] + [x for x in data[peer]["hosts"] if x != node]
+                    else:
+                        res[node].append(peer)
+        return res
+
+    def deep_merge_dicts(lhs=None, rhs=None):
+
+        local_lhs = lhs
+        local_rhs = rhs
+
+        if local_lhs is None or local_rhs is None:
+            return
+        for key, value in local_rhs.items():
+            if key in local_lhs:
+                local_lhs[key] = local_lhs[key] + value
+            else:
+                local_lhs[key] = value
+
+        return local_lhs
+
+    def generate_dot_syntax_from_dict(dict=None):
+
+        if dict is None:
+            return
+
+        res = ""
+
+        for label, nodes in dict.items():
+            for node in nodes:
+                res += '"{0}" -> "{1}";\n'.format(label, node)
+
+        return res
+
+    def detect_cycles(dict=None, data=None):
+
+        if dict is None:
+            return
+
+        for key, nodes in dict.items():
+            for node in nodes:
+                if "peers" in (data["_meta"]["hostvars"][node].keys()):
+                    if key in (data["_meta"]["hostvars"][node]["peers"]).split(","):
+                        raise Exception("Cycle Detected Between [{0}] <-> [{1}]".format(key, node))
+        return None
+
+    def assert_node_type(self, host=None, vars=None, group_name=None, valid_types=None):
+        """
+        Members of given group_name must have a valid node_type.
+        """
+        if 'node_type' not in vars.keys():
+            return valid_types[group_name]['default_type']
+
+        if vars['node_type'] not in valid_types[group_name]['types']:
+            raise AnsibleError(
+                'The host %s must have one of the valid node_types: %s' %
+                (host, ', '.join(str(_) for _ in list(valid_types[group_name]['types'])))
+            )
+        return vars['node_type']
+
+
+    def assert_unique_group(self, task_vars=None):
+        """
+        A given host cannot be part of the automationcontroller and execution_nodes group.
+        """
+        automation_group = task_vars.get('groups').get('automationcontroller')
```

Comment:
Defensively using `.get()` is a good idea, but chaining like this could wind up with `None` being the result of the first call and thus an error for the second.  We probably want `.get('groups', {}).get(...)`